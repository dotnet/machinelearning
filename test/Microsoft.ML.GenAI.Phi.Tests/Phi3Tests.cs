// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Text;
using System.Text.Json;
using ApprovalTests;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using FluentAssertions;
using Microsoft.ML.GenAI.Core.Extension;
using TorchSharp;
using Xunit;

namespace Microsoft.ML.GenAI.Phi.Tests;

[Collection("NoParallelization")]
public class Phi3Tests
{
    public Phi3Tests()
    {
        if (Environment.GetEnvironmentVariable("HELIX_CORRELATION_ID") != null)
        {
            Approvals.UseAssemblyLocationForApprovedFiles();
        }

        torch.set_default_device("meta");
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Mini4KShapeTest()
    {
        var model = new Phi3ForCasualLM(Phi3Config.Phi3Mini4kInstruct);
        var stateDictStr = model.PeekShape();
        Approvals.Verify(stateDictStr);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Mini4KInt8QuantizeShapeTest()
    {
        var model = new Phi3ForCasualLM(Phi3Config.Phi3Mini4kInstruct);
        model.ToInt8QuantizeModule();
        var size = model.GetSizeInBytes();
        var stateDictStr = model.PeekShape();
        var sizeInGB = size / 1024 / 1024 / 1024;
        sizeInGB.Should().Be(3);
        Approvals.Verify(stateDictStr);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Mini4KInt4QuantizeShapeTest()
    {
        var model = new Phi3ForCasualLM(Phi3Config.Phi3Mini4kInstruct);
        model.ToInt4QuantizeModule();
        var size = model.GetSizeInBytes();
        var stateDictStr = model.PeekShape();
        var sizeInGB = size / 1024 / 1024 / 1024;
        sizeInGB.Should().Be(2);
        Approvals.Verify(stateDictStr);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Medium4KShapeTest()
    {
        var model = new Phi3ForCasualLM(Phi3Config.Phi3Medium4kInstruct);
        var stateDictStr = model.PeekShape();
        Approvals.Verify(stateDictStr);
    }


    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Medium128KShapeTest()
    {
        var model = new Phi3ForCasualLM(Phi3Config.Phi3Medium128kInstruct);
        var stateDictStr = model.PeekShape();
        Approvals.Verify(stateDictStr);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Mini128KShapeTest()
    {
        var model = new Phi3ForCasualLM(Phi3Config.Phi3Mini128kInstruct);
        var stateDictStr = model.PeekShape();
        Approvals.Verify(stateDictStr);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Mini128KLayerSizeTest()
    {
        var model = new Phi3ForCasualLM(Phi3Config.Phi3Mini128kInstruct);
        var size = model.GetSizeForEachDynamicLayerInBytes();
        // convert size to MB
        var sizeInMB = size.ToDictionary(x => x.Key, x => x.Value / 1024 / 1024);

        var json = JsonSerializer.Serialize(sizeInMB, new JsonSerializerOptions { WriteIndented = true });
        Approvals.Verify(json);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void TokenizerTest()
    {
        var modelWeightFolder = Path.Join("Llama");
        var modelPath = Path.Join(modelWeightFolder, "tokenizer.model");
        var tokenizer = Phi3TokenizerHelper.FromPretrained(modelPath);
        tokenizer.BeginningOfSentenceId.Should().Be(1);
        tokenizer.EndOfSentenceId.Should().Be(2);

        // test <|end|>
        var endIds = tokenizer.EncodeToIds("<|end|>", addBeginningOfSentence: false, addEndOfSentence: false, considerPreTokenization: false, considerNormalization: false);
        endIds.Should().BeEquivalentTo(new int[] { 32007 });

        var messages = new string[]
        {
            "Can you provide ways to eat combinations of bananas and dragonfruits?",
            "Sure! Here are some ways to eat bananas and dragonfruits together: 1. Banana and dragonfruit smoothie: Blend bananas and dragonfruits together with some milk and honey. 2. Banana and dragonfruit salad: Mix sliced bananas and dragonfruits together with some lemon juice and honey.",
            "What about solving an 2x + 3 = 7 equation?",
            "\nCount to 3\n",
            "<|user|>",
            "<|end|>",
            "<|assistant|>",
            "<|user|>\nCount to 3<|end|>\n<|assistant|>",
        };
        var sb = new StringBuilder();
        foreach (var message in messages)
        {
            var tokenizeIds = tokenizer.EncodeToIds(message, true, false, considerPreTokenization: true);
            var decodeToString = tokenizer.Decode(tokenizeIds, considerSpecialTokens: true);
            sb.AppendLine(decodeToString);
            var tokenizedStr = string.Join(", ", tokenizeIds.Select(x => x.ToString()));

            sb.AppendLine(tokenizedStr);
        }
        Approvals.Verify(sb.ToString());
    }
}
