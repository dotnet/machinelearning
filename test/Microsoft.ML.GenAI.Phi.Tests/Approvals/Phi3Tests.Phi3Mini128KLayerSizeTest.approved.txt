{
  "model.layers.0": 216,
  "model.layers.1": 216,
  "model.layers.2": 216,
  "model.layers.3": 216,
  "model.layers.4": 216,
  "model.layers.5": 216,
  "model.layers.6": 216,
  "model.layers.7": 216,
  "model.layers.8": 216,
  "model.layers.9": 216,
  "model.layers.10": 216,
  "model.layers.11": 216,
  "model.layers.12": 216,
  "model.layers.13": 216,
  "model.layers.14": 216,
  "model.layers.15": 216,
  "model.layers.16": 216,
  "model.layers.17": 216,
  "model.layers.18": 216,
  "model.layers.19": 216,
  "model.layers.20": 216,
  "model.layers.21": 216,
  "model.layers.22": 216,
  "model.layers.23": 216,
  "model.layers.24": 216,
  "model.layers.25": 216,
  "model.layers.26": 216,
  "model.layers.27": 216,
  "model.layers.28": 216,
  "model.layers.29": 216,
  "model.layers.30": 216,
  "model.layers.31": 216
}